runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java

```diff
@@ -466,36 +495,55 @@ private static void rejectSplittable(DoFn<?, ?> doFn) {
 
   private static class ParDoTranslatorBatch<InputT, OutputT>
       implements FlinkBatchPipelineTranslator.BatchTransformTranslator<
-      ParDo.MultiOutput<InputT, OutputT>> {
+      PTransform<PCollection<InputT>, PCollectionTuple>> {
 
     @Override
     @SuppressWarnings("unchecked")
     public void translateNode(
-        ParDo.MultiOutput<InputT, OutputT> transform,
+        PTransform<PCollection<InputT>, PCollectionTuple> transform,
         FlinkBatchTranslationContext context) {
-      DoFn<InputT, OutputT> doFn = transform.getFn();
+      DoFn<InputT, OutputT> doFn;
+      try {
+        doFn = (DoFn<InputT, OutputT>) ParDoTranslation.getDoFn(context.getCurrentTransform());
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
       rejectSplittable(doFn);
       DataSet<WindowedValue<InputT>> inputDataSet =
           context.getInputDataSet(context.getInput(transform));
 
       Map<TupleTag<?>, PValue> outputs = context.getOutputs(transform);
 
+      TupleTag<?> mainOutputTag;
+      try {
+        mainOutputTag = ParDoTranslation.getMainOutputTag(context.getCurrentTransform());
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
       Map<TupleTag<?>, Integer> outputMap = Maps.newHashMap();
       // put the main output at index 0, FlinkMultiOutputDoFnFunction  expects this
-      outputMap.put(transform.getMainOutputTag(), 0);
+      outputMap.put(mainOutputTag, 0);
       int count = 1;
       for (TupleTag<?> tag : outputs.keySet()) {
         if (!outputMap.containsKey(tag)) {
           outputMap.put(tag, count++);
         }
       }
 
+      // TODO: Why does the UnionCoder order have to match the output map order? Why does this
```

Comment:
The map of tags is created here: https://github.com/apache/beam/blob/875f3e1b0c943b56bda99e610ede872a07573be5/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java#L487

And it's used again here to create individual Flink `DataSet`s for each of the output tag indices: https://github.com/apache/beam/blob/875f3e1b0c943b56bda99e610ede872a07573be5/runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java#L571

If the order, i.e. the index, changes in between then the mapping to outputs won't be correct anymore.

Reply:
That first link is exactly what's happening above. Unfortunately, it's insufficient in getting the union coder to work properly. Leaving that as-is breaks when I do a protobuf round-trip. I have to create a union coder that lists individual coders in the same order they appear in the output map in order to get tests to pass. My question is: why is this only necessary when using a rehydrated pipeline?

Reply:
I think it might be because rehydration messes up the order of the individual coders. I also just realised that the code that constructs the `outputMap` is making sure to put the main coder at index `0` while the code that constructs the lists of union coders doesn't do it. I think this just happened to work because the main-output coder always was at index `0`.

Reply:
To be more specific: I think rehydration changes the order so that the main input is no longer at index `0`.

Reply:
Ah, this is interesting. Is there any requirement that the rest of the outputs match the order of their respective coders?

Reply:
UnionCoder requires that the order of the inputs / outputs for the tags to match because the union coder encodes values in a specific order and when reading them in needs to decode them in that same order.

Reply:
OK, thanks for confirming this. I'll leave the fix as-is and remove the comment.